compositor_pipeline/src/pipeline/whip_whep.rs

```diff
@@ -0,0 +1,202 @@
+use crate::pipeline::input::whip::depayloader::Depayloader;
+use axum::{
+    routing::{delete, get, patch, post},
+    Router,
+};
+use compositor_render::InputId;
+use error::WhipServerError;
+use handlers::{handle_whip, status, terminate_whip_session, whip_ice_candidates_handler};
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    sync::{Arc, Mutex, Weak},
+    time::{Duration, Instant},
+};
+use tokio::sync::mpsc;
+use tower_http::cors::CorsLayer;
+use tracing::{error, warn};
+use webrtc::{
+    api::{
+        interceptor_registry::register_default_interceptors,
+        media_engine::{MediaEngine, MIME_TYPE_H264, MIME_TYPE_OPUS},
+        APIBuilder,
+    },
+    ice_transport::ice_server::RTCIceServer,
+    interceptor::registry::Registry,
+    peer_connection::{configuration::RTCConfiguration, RTCPeerConnection},
+    rtp_transceiver::{
+        rtp_codec::{RTCRtpCodecCapability, RTCRtpCodecParameters, RTPCodecType},
+        rtp_transceiver_direction::RTCRtpTransceiverDirection,
+        RTCRtpTransceiverInit,
+    },
+};
+
+mod error;
+mod handlers;
+mod helpers;
+
+use crate::{queue::PipelineEvent, Pipeline};
+
+use super::EncodedChunk;
+
+pub async fn run_whip_whep_server(pipeline: Weak<Mutex<Pipeline>>) {
+    let pipeline_ctx = match pipeline.upgrade() {
+        Some(pipeline) => pipeline.lock().unwrap().ctx.clone(),
+        None => {
+            warn!("Pipeline stopped.");
+            return;
+        }
+    };
+
+    if !pipeline_ctx.start_whip_whep {
+        return;
+    }
+
+    let state = pipeline_ctx.whip_whep_state;
+    let port = pipeline_ctx.whip_whep_server_port;
+
+    let app = Router::new()
+        .route("/status", get(status))
+        .route("/whip/:id", post(handle_whip))
+        .route("/session/:id", patch(whip_ice_candidates_handler))
```

Comment:
what does it mean to handle ice candiadates? I suspect `whip::new_ice_candidates_handle` would be better and rename handlers to whip_handlers maybe